### ISSUE TITLE
add PiP support

### ISSUE THREAD

**Issue by BO41** [Chocobozzz/PeerTube#1211]
I think it would be nice to support picture in picture, just like youtube now, or even better, like twitch.tv

**What happened?**

The video stops playing when I switch to a different site, or look for another video.
And I am unable to see the video when I scroll down to read a comment.

**What do you expect to happen instead?**

A small player should pop up and show the video.

parts:
- [ ] button to pop out pip for changing site https://github.com/Chocobozzz/PeerTube/issues/1211#issuecomment-427708393
- [x] open pip on scroll down
- [x] add video controls to pip player
- [ ] add options to control pip behavior

**Comment by rigelk** [comment]
I imagine it like the video miniature player that goes to the corner in YouTube's mobile app. Is it how you envisioned it?

However it would only work while you stay on the same site, not beyond when you switch to a different site.

**Comment by BO41** [comment]
@rigelk yes, exactly. Like the youtube app, youtube.com, newpipe, twitch.tv, ...
And of course only on PeerTube and the Angular app.

![screenshot_2018-10-06 twitch](https://user-images.githubusercontent.com/29450609/46570493-5208fa00-c965-11e8-9e2e-165817a750d8.png)
twitch.tv if you scroll or click somewhere else
![screenshot_2018-10-06 youtube](https://user-images.githubusercontent.com/29450609/46570497-59300800-c965-11e8-869e-e5b6b8902f91.png)
youtube if you click on the PiP button

**Comment by rigelk** [comment]
Thanks for the screenshot! That's what I was missing. I didn't know they had a corner player (PiP sounds more like putting a video in a corner of the video itself), but it looks great.

**Comment by McFlat** [comment]
+1

**Comment by BO41** [comment]
@XenonFiber I see. Good point :+1: 
And having the small window when scrolling? This shouldn't bother, right?

**Comment by XenonFiber** [comment]
@BO41 That would probably be fine since you're still on the same video page. It feels like something you should still have to choice to disable, though.

**Comment by rigelk** [comment]
Following PR #1345, PR #1463 takes over. It's still needs to be disabled by default, and expose the option to enable it to non-authenticated users.

**Comment by BenLubar** [comment]
You can right click a video, then right click again without moving the mouse, and then select "Picture in Picture" from the browser menu in browsers that support it.

This method also works on YouTube.

**Comment by BO41** [comment]
Most major browsers now support PiP, so no need to implement it

**Comment by Chocobozzz** [comment]
I think we can keep this issue opened, because for now when you move to another page you loose the player. Maybe we can keep the native pip player when we change page since we are in a SPA but I don't really read the pip specs.

**Comment by BenLubar** [comment]
All you need to do to keep the native PiP player working on page change is don't delete the HTML `<video>` element and reuse it instead

Hide it and pause it when you're not on a page that plays a video, but don't delete it.

**Comment by kimsible** [comment]
Hi @rigelk would you have a current work on PiP or did you plan to push this feature later in 2021 ?

**Comment by rigelk** [comment]
@kimsible it is not on my todo in the foreseeable future, nor do I have pending work on this.

**Comment by kimsible** [comment]
Thank you @rigelk for the details. Is there anything I can do like extracting the videojs wrapper to make the native PiP work with the navigation left menu ?

**Comment by rigelk** [comment]
@kimsible it depends what you are aiming for, and this is potentially a huge refactoring ahead. But why the navigation left menu?

**Comment by kimsible** [comment]
@rigelk actually all navigation, once the videojs component is destroyed with a new route. We may extract the videojs-wrapper outside the router-outlet. Its seem to be a huge refactoring, but maybe there is a way with angular.

**Comment by kimsible** [comment]
After checking, all mainstream video platform, only Dailymotion have a full support : 

- YouTube : enabled native PiP without navigation support / home made mini player with navigation support
- Facebook : disabled native PiP / home made mini player with navigation support
- Vimeo : disabled native PiP / home made mini player with the navigation support
- Dailymotion : **enabled native PiP with navigation support** / home made mini player with navigation support

The player wrapper of Dailymotion is indeed placed above the main root component with absolute css position.

My knowlegde of Angular seems a little bit limited to this kind of refactor but we may try to display the `<my-video-watch></my-video-watch` component above the `<router-outlet></router-outlet>` in the app main component template.

The `video-watch` component should only display its main view when the router is on the videos/watch route. Videojs instance should be destroyed only when routing and native PiP is not used.

In any case, I don't think Firefox has an API like Chrome to know is the native PiP is used, so we must have our own home made mini player like Dailymotion does while Firefox does not support the official W3C API.

What do you think ?

**Comment by Casuallynoted** [comment]
> This is a nice feature, but I would prefer how YouTube handles it with a button over how Twitch automatically switches to it. I don't think websites should break continuity; if I'm clicking away from a video, I don't expect the video to come with (unless I explicitly wanted it to, of course).

Would there possibly be a way to override that for, for example, music and podcasts? It feels especially natural to have it for stuff like that where an interruption in the music feels abrupt and, say, if you want to browse the rest of peertube while listening.
